Running tests on Windows

cderv opened the pull request:
**WIP: Draft PR until it is working** 

- [x] Concert `run-test.sh` to a `run-test.ps1` version
- [x] `./configure.cmd` should run on CI - adapt our custom `quarto-dev` action
- [x] Deal with python at as simply as possible to make this works 
- [x] Deal with Julia environment (new tests added in #4012)
- [x] Adapt our actions to handle windows OS too 
- [x] Fix `renv` activation when running tests ([issue](https://github.com/quarto-dev/quarto-cli/actions/runs/4017804431/jobs/6902627415#step:18:37))
- [x] Improve renv caching to respect R version
- [x] Fix failed tests (possibly in the PR or open issue for each one found)
	- [x] https://github.com/quarto-dev/quarto-cli/issues/4124 
	- [x] https://github.com/quarto-dev/quarto-cli/pull/4255
	- [x] https://github.com/quarto-dev/quarto-cli/issues/4260
	- [x] https://github.com/quarto-dev/quarto-cli/issues/4337
	- [x] https://github.com/quarto-dev/quarto-cli/pull/4345
- [x] Adapt PR to recent changes to workflows on #4354 and following commits https://github.com/quarto-dev/quarto-cli/compare/19fcc97946387fa18bee27454e81a72c5b8746b0...8c538d3f574bc167ca2325de4a553b0b2f669b03
- [x] Improve the python environment setup as this is not well designed for local and CI usage currently. 
- [x] Improve the julia environment setup to be like Python and R 

## Follow up work 

- https://github.com/quarto-dev/quarto-cli/issues/4339

The following are related to ignored tests on Windows and should be fixed
- https://github.com/quarto-dev/quarto-cli/issues/4347
- https://github.com/quarto-dev/quarto-cli/issues/4194

## Adaptations 

-  Removing `./tests/bin/*` folder 
	- 	This folder and  `source bin/activate` were not useful (some hardcoded miniconda path in it). Removing it seems to work ok. 
-  Adding a `.python_version` file to pin a python version to use in `tests/`
	- 	This is understood by `setup-python`
	- 	It allows to use `pyenv` to automically use the right version of python in when in `tests/`
- Supporting Windows in our custom action `quarto-dev` which install quarto development version
	- It was not an easy piece to make things work on Unix and Windows. Especially because we run a `quarto.cmd`, which is not found in PATH on windows's `bash`. Using `cmd` shell is a pain as it seems you can't really set `%GITHUB_PATH%` on GHA (🤷‍♂️ ). So for now, it is hardcoded path but won't probably change soon. Anyway it is working now. 
- Adding some caching for Julia plus adaptation so that it runs on Windows too. 

## Probably some more to change 

* We are setting up environment in GHA in some steps, but we are also doing `renv::restore()` and `pip install` in the `run-test`. It seems a duplication. I am thinking of separating `config-tests-env` script for configure a test environment, and then having `run-test` to just run the test. But still need to think through CI usage vs Local usage.

* Having a `quarto.ps1` instead of `quarto.cmd` would be better for dev windows user. I never used CMD, and other windows dev probably use either Powershell or Git Bash. 

## Notes on developing tests for Linux and windows 

* When using regex matching, newline needs to be OS independant (345294069040fc3e26662c2d73f971f585157f6c)
* If a tests does not need to be run on several OS (too costly to adapt, or not necessary to make OS independant), it can be easily ignored (418d0776dfc1d644a5025119eafd33cdd53fe255)
* Some calls to `quarto` needs to be explicitly made to `quarto.cmd` on Windows because in dev version `quarto.exe` is not available and `quarto` without extension will not be found (6be823273c7bd666e7d18f2d8caa8fdca0a257b7)
* Windows file locking system must be taken into account (example in cleaning where one need to move out a directory before removing it c29a182c3a8d596567074b2425cf8e95df760770)
* Windows paths specificity needs to be dealt with (example with path ending `/` vs `\` 9f37289bf945f6410cd4a3051cd96d6bdc29d0a4)

_More comments to come as I go through with it_

cderv committed:
First version of powershell script without python running

cderv committed:
Setup python environment for Windows and UNIX

cderv committed:
update python package

cderv committed:
Try something for python on windows

- Be sure to deactivate Quarto env var that forces PYTHON to overwrite local config
- Use `py.exe` because in `venv` the `exe` do not work somehow.

cderv committed:
Activate tests on Windows

cderv committed:
Adapt our action for windows

cderv committed:
Run effectively CI on windows

cderv committed:
Try unix with the python virtualenv setup that probably is not used at all

cderv committed:
Forgot ;; in the bash case syntax

cderv committed:
Fix all syntax

cderv committed:
.python-version is not found as in tests/ and not root project

cderv committed:
Separate in one step per OS

dragonstyle commented:
I would like to note that this is an extremely exciting PR.

cderv committed:
Put back setup in main workflow

We duplicate the step in the `run-test` script, but it is better to have this setup before installing quarto-dev in next step

cderv committed:
We need some platform specifier in requirement.txt

cderv committed:
[debug] Adapt GHA to debug

* Using tmate
* Deactivating other workflow as not useful when debugging

cderv committed:
Fix configure.cmd to work on CI

cderv committed:
[debug] deactivate tmate

cderv committed:
Correctly put quarto to PATH for Unix and Windows

cderv committed:
Merge branch 'main' into windows/run-tests

cderv committed:
Check that new Julia addition is also working

cderv committed:
Correct quote for Windows and UNIX

cderv committed:
Fix merge problem in figute test for Julia

cderv committed:
fIx custom action to work on  Windows and UNIX

cderv committed:
Add caching to Julia

cderv committed:
Try with no quote to export an ENV in github action with cmd shell

cderv committed:
Correct action for Julia Cache

cderv committed:
Try with powershell instead of cmd for window sheel in custom action

cderv committed:
Try using pwsh shell available for all OS too

as on Windows it will match `quarto` in PATH for `quarto.cmd` which does not happen in bash shell

cderv committed:
Show PATH to check it is correctly set

cderv committed:
[debug] try tmate in composition action

cderv committed:
Correct test for sanity check

cderv committed:
[debug] Only debug windows now

cderv committed:
[debug] Only run windows now

cderv committed:
we need to stay in CMD to access variable

cderv committed:
Try to access variable again in CMD

cderv committed:
Remove the double backslash on windows

cderv committed:
fixup: Remove the double backslash on windows

cderv committed:
Retry with Powershell and set PATH hardcoded in action

cderv committed:
[debug] Move tmate to debug Julia installation

cderv committed:
Keep original command for julia but enforce bash as shell

cderv committed:
[debug] Add tmate later to debug renv

cderv committed:
For multi OS we need another RENV_PATH_ROOT

cderv committed:
Set RENV_PATHS_ROOT later

cderv committed:
[debug] Check on linux and no tmate

cderv committed:
Add a configure helper script

cderv committed:
Merge branch 'main' into windows/run-tests

cderv committed:
[tests-fix] Handle windows path correctly to detect folder

cderv committed:
[tests-fix] On windows, quarto.cmd needs to be run, not just quarto

cderv committed:
Fix TMPDIR on windows to avoid different disk drives

cderv committed:
Merge branch 'main' into windows/run-tests

cderv committed:
Fix tempdir for Deno to avoid different drive issue

On Windows the default / system temp directory is on `C:/` whild projects and all are on `D:/` - this creates issues when moving file around (like we do with Deno)

cderv committed:
[bug-fix] Check first for http(s) protocol before testing local

On Window, `existsSync()` on a URL path does not work as it does on Linux. So filtering those urls

fix #4124 - Installing extension from url fails on Windows

cderv committed:
[tests-fix] Go back to parent dir before removing current directory

It seems to work on Linux, but not on windows were a "lock as in use" issue is thrown

cderv committed:
quarto.cmd needs to be explicit on Windows

cderv committed:
test better for http(s) remote extension

cderv committed:
Revert "test better for http(s) remote extension"

This reverts commit e4dfa93db1664a3a80a6cd8dfbb79858446d6605.

dragonstyle committed:
Merge pull request #4129 from quarto-dev/windows/fix-4124

Check first for http(s) protocol before testing local

cderv committed:
Merge branch 'windows/run-tests' of https://github.com/quarto-dev/quarto-cli into windows/run-tests

cderv committed:
Merge branch 'main' into windows/run-tests

cderv committed:
Merge branch 'windows/run-tests' of https://github.com/quarto-dev/quarto-cli into windows/run-tests

cderv committed:
Merge branch 'main' into windows/run-tests

cderv committed:
Add Julia configuration

cderv committed:
Checking newline should be OS independant

cderv committed:
Allow to ignore tests

Some tests can be run only on one OS for example, and porting to Windows is not worth it

cderv committed:
Add a  configuration for VSCODE to debug in tests

Current configuration debug when `deno run`. Here we want to debug when `deno test`

cderv committed:
Do not error on file existence when filtering resources (#4255)

cderv committed:
Merge branch 'main' into windows/run-tests

cderv committed:
Merge branch 'main' into windows/run-tests

cderv committed:
Show path to test as relative to tests directory.

`expandGlob*` was returning absolute path which lead to long absolute path in tests log preventing easy copy of failing tests

cderv committed:
[fix] Correctly handle windows file path on windows for test logging

cderv committed:
[fix] Ignore qualified path tests on WIndows

This is temporary waiting for #4339

cderv committed:
Merge branch 'main' into windows/run-tests

cderv committed:
Merge branch 'main' into windows/run-tests

cderv committed:
Merge branch 'main' into windows/run-tests

cderv commented:
@cscheid I see you have reorganized the workflow file following addition of performance checks. I added an item in my todo list above to adapt. 
Do we we want to run performance checks on Windows too ?

cderv committed:
Merge branch 'main' into windows/run-tests

cderv committed:
[fix] Ignore test rendering to stdout for now

cderv committed:
[gha] Invalidate R cache based on OS and R version

cderv committed:
[test workflow] same name as linux script

cderv committed:
[fix] Fix playwright test by using quarto.cmd on Windows

cderv committed:
[gha] Use pak to install dependency

As curl was updated in lock file and require missing system deps

cderv committed:
[gha] Fix name of script in worflow

cderv committed:
[gha] Correct env var

cderv committed:
[gha] correct relative path

cderv committed:
[fix] Use `npx.cmd` on Windows

cderv committed:
[test environment] handle tinytex download failure due to API rate limitation

cderv committed:
[test environment] Use pipenv to handle python environment

This will be better for several reasons

- A manifest file to track our required dependencies
- Handling several type of dependencies environment, like last versions of packages or lock version
- Handling better virtualenvironment for local project on Windows and Linux
- Allow to install to system python also

Only requirement : having pipenv installed

cderv committed:
[test environment] Add the Julia Manifest of locked packages

cderv committed:
Update helper configuration script

cderv committed:
[gha] Fix linux by installing missing curl deps

cderv committed:
[gha] PAK does not work for renv::restore() yet

cderv committed:
[test environment] pexpect module is required for quarto check on Linux

It is not needed on Windows it seems

cderv committed:
Merge branch 'windows/run-tests' of https://github.com/quarto-dev/quarto-cli into windows/run-tests

cderv committed:
[test environment] update configuration file

cderv committed:
[test environment] Update python lock file on windows

cderv committed:
[test environment] Also check and configure R environment

cderv committed:
[test environment] Let install pipenv for the developer

cderv committed:
[test environment] script should be executable

cderv committed:
[test environment] Add logging to configure script

cderv committed:
[test environment] ptyprocess seems required on Linux in GHA

cderv committed:
[test environment] Windows adjustement in pipfile.lock

cderv committed:
[gha] Clean workflow and reactivate other workflows

cderv committed:
[test envrionment] Initiate a README for tests

[skip ci]

cderv committed:
[test environment] Auto activate the virtualenv when running tests

cderv committed:
[test environment] Run configuration as part of running tests

cderv committed:
[test environment] Improve documentation in README

cderv committed:
Merge branch 'main' into windows/run-tests

cderv commented:
This is now ready for integration. I just launched a last test with the latest change from main. 

All tests passes except 2 or 3 I think that I ignore for now until we fixe them. There are quite a few changes for local environment so it is worth reading and checking before we merge. A [README ](https://github.com/quarto-dev/quarto-cli/blob/windows/run-tests/tests/README.md)is available to explain. 

This PR 

- Revisit how we set up our testing environment locally (for R Python Julia) - Currently it is with locked state we'll update on the go. Challenge was to have something working for multi OS seemlessly; Using pipenv for Python helps with that
- Adapte run-tests.sh  and add a run-tests.ps1
- Adapt GHA workflow to run on windows
- Fix some tests
- Fix some bugs in code (and other fix were done in external PR)
- If anyone is available tomorrow for a quick demo and presentation, and to test the new local configuration for test environment that would be cool.

Hopefully the change will work for everyone local environment. 

## Limitation and what to do next

Currently we rely on testing our code with locked state of packages for Julia, R and Python. This is using tools that will rely on lockfile to recreate a fix state. 

This means that we'll need to update the different lockfile on regular basis - I should add a note in the README probably

Making it possible to test with latest version of all packages will be next in line among test environment improvment.

cderv commented:
@dragonstyle putting you as reviewer as you seemed to be available and interested to test all this locally together.

cderv committed:
[test environment] Don't use which to find if a command is available

cderv committed:
[gha] Don't run playwright tests on windows

this will save us some time on workflow

cderv committed:
[test environment] Be more clear on what check if something is missing

cderv committed:
[test environment] Update README.md with more info on requirements

[skip ci]

cderv committed:
[fix]  Move the exception at upper level

cscheid commented on tests/run-tests.sh:
```diff
@@ -26,18 +26,19 @@ export QUARTO_DEBUG=true
-# Ensure that tinytex is installed
-quarto install tinytex --no-prompt
+# Activating python virtualenv
+if [[ -z $VIRTUAL_ENV ]]
```
```
if [[ -z $FORCE_NO_PIPENV ]]
then
  OLD_VIRTUAL_ENV=$VIRTUAL_ENV
  if [[ -n $VIRTUAL_ENV ]]
  then
    deactivate
  end
  echo "> Activating virtualenv for Python tests"
  source "$(pipenv --venv)/bin/activate"
  quarto_venv_activated="true"
end
# ...
# at the end
# this is probably not correct if we were not in a venv but you get the idea.
if [[ -n $OLD_VIRTUAL_ENV ]]
then
  deactivate
  source $OLD_VIRTUAL_ENV/bin/activate
end

```

cderv replied:
So I did another approach as deactivating was not robust. For example I am using `pyenv` on Linux to manage my python and virtualenv  and `deactivate` was not existing. 

So I leveraged some **pipenv** feature. Here is how it works

* `pipenv` will always create a virtual environment even when run in another virtuaenv. This means by default it will activate the new virtualenv. This environment will be located in  `tests/.venv`
* If I detect that $VIRTUAL_ENV is set though before activating with pipenv, I remember this, and at the end, I run `source $OLD_VIRTUAL_ENV/bin/activate` to reactivate the initial virtualenvironment. 
* To skip the default behavior `QUARTO_TESTS_FORCE_NO_PIPENV` can be set to any value. In that case, tests are running in the current python environment with no change. 

I believe this work well and answer our discussion.

cderv commented:
Thanks @cscheid and @dragonstyle - I'll still do the last adjustment discussed today, and this will be ready to merge tomorrow ! 🎉 

So cool !

cderv committed:
[test environment] Add a mechanism to work within an existing python virtualenv shell

This relies on pipenv features to ignore virtualenv and create a new one.

Add a mechanism to skip using pipenv

cderv committed:
[test environment] Update documentation in README

[skip ci]

cderv commented:
I did the modification for the virtual environment, and added some information to the documentation. Ready now to merge !

cscheid commented:
Feel free to merge anytime!

cderv committed:
Revert "[gha] Don't run playwright tests on windows"

This reverts commit 3e66c4eee5fbe81ba4dfceb6e5b2c55d07631f57.

cderv committed:
[gha] Run playwright on Windows at least once a day by scheduling workflow
